### PR TITLE
Feature/priority-queue/

### PR DESCRIPTION
**Pull request by SaxanLosterToreda**

### CODE REVIEW

File: src/priority-queue.ts [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+		return false;
+	}
+	if (typeof nodeTwoIndex !== 'number') {
+		return false;
+	}
+
+	return JSON.stringify(this.elements[nodeOneIndex]) < JSON.stringify(this.elements[nodeTwoIndex]);
+};
+
+export class ArmorPriorityQueue<T> implements ArmorCollection<T> {
```
Comment (mbrich) [inline src/priority-queue.ts]: Make ArmorPriorityQueue<T> the file's default export:
`export default class ArmorPriorityQueue<T>`

The class matching the filename should be the default export and all other classes (if any) should be regular exports.

File: src/priority-queue.ts [hunk 2]
```diff
@@ -0,0 +1,262 @@
+import {ArmorCollection} from './collection';
+import {ArmorCollectionSelector} from './selector';
+
+/**
+ * decide what to do when no comparator is passed
+ */
+
+export interface ArmorPriorityQueueOptions<T> {
```
Comment (mbrich) [inline src/priority-queue.ts]: Move ArmorPriorityQueueOptions<T> to a separate file with name `priority-queue-options`. It should also be the default export.
While not a big deal here, separating classes into separate files is a beest practice as it improves _[tree shaking](https://webpack.js.org/guides/tree-shaking/)_ during the webpacking process.

File: src/priority-queue.ts [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+	public push(element: T): ArmorPriorityQueue<T> {
+		this.elements.push(element);
+		this.fixHeap(this.size() - 1);
+		return this;
+	}
+
+	public pop(): T | null {
+		if (this.size() === 0) {
+			return null;
+		}
```
Comment (mbrich) [inline src/priority-queue.ts:296]: General note: adding newlines between early drops is helpful when scrolling through. Not a strict rule but at your discretion. I recommend a newline between early drops if you have 1 or 2 at the start of a function.

File: src/priority-queue.ts [hunk 4]
```diff
@@ -10,6 +10,17 @@ export interface ArmorPriorityQueueOptions<T> {
 	comparator?: (nodeOneIndex: number, nodeTwoIndex: number) => boolean;
 }
 
+const defaultComparator = function (nodeOneIndex: number, nodeTwoIndex: number) {
```
Comment (mbrich) [inline src/priority-queue.ts]: Users need a function signature to match to write their own comparator. 
Create a new file for the comparator signature called `priority-queue-comparator` that looks roughly like this:
```typescript
export default interface ArmorPriorityQueueComparator<T> {
   (a: T, B: T) => boolean;
}
```

Comment (mbrich) [inline src/priority-queue.ts]: - [ ] Import the comparator into `priority-queue` and use it in the defaultComparator declaration. 
Why: Defining comparator type will enable strict type checking against what it should be.

- [ ] Change arguments a and b from an index to a generic type <T>. 
Why: Using an index requires comparator to be responsible for retrieving elements from the array. Remember that users will be writing code here and an array out of bounds error when ndx math is wrong. Comparators have one job: compare two things and return the result based on some criteria.
By changing a and b to generic types, the function and caller no longer need internal class implementation knowledge. If I write my own comparator the signature tells me I get two objects, and do something to compare.

```typescript
const defaultComparator: ArmorPriorityQueueComparator<T> = (a: T, b: T) => {
   ...
}
```

File: src/priority-queue.ts [hunk 5]
```diff
@@ -10,6 +10,17 @@ export interface ArmorPriorityQueueOptions<T> {
...
+	if (typeof nodeOneIndex !== 'number') {
+		return false;
+	}
+	if (typeof nodeTwoIndex !== 'number') {
+		return false;
+	}
+
+	return JSON.stringify(this.elements[nodeOneIndex]) < JSON.stringify(this.elements[nodeTwoIndex]);
```
Comment (mbrich) [inline src/priority-queue.ts]: - [x] Remove the default comparator or move to the test file.

Two issues to highlight here. 
1) `JSON.stringify` serializes an input and returns the result as a string.
**Why:** Array elements are ordered. Dictionaries and maps are not. Object order is not guaranteed and not predictable. Object properties may frequently appear in a specific order but are also not guaranteed to always appear in the same order. This fact is especially important when testing equality with the identity operator `===`, because same exact object can potentially serialize as two different strings due to ordering. When this happens comparing with `===` will produce false, even though the original object is equal.

2) Comparing strings using the `<`, `>`, or `<=`, `>=` operators will almost certainly not do what you're expecting. 
**Why:** [comparison operators compare strings lexicographically](https://stackoverflow.com/questions/10863092/why-is-string-11-less-than-string-3). Usually creates unintended results, as most comparisons you would want to do, e.g. is a shorter length than b, don't make sense with a lexicographical comparison. Additionally, string casing also affects the result.

Both 1 & 2 result in indeterminate behavior.

Lastly, you can actually remove the default comparator entirely. If used in your unit tests, move the comparator there. Because the object stored in the queue is generic type <T>, there is no way to know what comparison operator will produce a valid result. For examples, numbers will be ordered logically, but strings won't be.

File: src/priority-queue.ts [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+};
+
+export class ArmorPriorityQueue<T> implements ArmorCollection<T> {
+	public elements: T[];
+	public comparator: (nodeOneIndex: number, nodeTwoIndex: number) => boolean;
```
Comment (mbrich) [inline src/priority-queue.ts]: - [x] Make it readonly

- [x] Change from inline function type to `ArmorPriorityQueueComparator<T>`

File: src/priority-queue.ts [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+
+	constructor(options: ArmorPriorityQueueOptions<T> = {}) {
```
Comment (mbrich) [inline src/priority-queue.ts]: - [ ] Change comparator type check into a validation of comparator type and throw when it fails (shown below).

Good use of backup default to fallback when comparator is not provided. Due to the complexities of generic types we can't reliably provide a default. Therefore, a user provided comparator is required and we can't do our job without one and it makes sense to throw if no comparator is provided. 

This:
```typescript
		if (typeof options.comparator === 'function') {
			this.comparator = options.comparator;
		} else {
			this.comparator = defaultComparator;
		}
```'
can be condensed down to a throw:
```typescript
if (typeof options.comparator !== 'function') {
   throw new Error('comparator error message here');
}
```

Comment (mbrich) [inline src/priority-queue.ts]: The options object provides a great way for users to optionally add parameters that modify behavior without requiring it. However, it's not ideal for required parameters unless we need to pass in a large set of required properties. General best practice (with exceptions) is to place required parameters in the constructor, with optional options last.

- [x] Remove comparator from options object.

- [x] Add required comparator as first constructor argument.

Comment (mbrich) [inline src/priority-queue.ts]: - [ ] Use the `?` operator to make options optional instead of defaulting to empty object.
```typescript
constructor(options?: ArmorPriorityQueueOptions<T>) {
  ...
}
```

Using a method to parse optional properties from an options/config object is helpful for testing. Testing whether the options object was passed in at all is an easy way to early drop from that method. To do that, the default case for options is undefined. 
Because the default is an empty object, the following is true:
```typescript
const options = {};

// empty objects are truthy.
// This will print 'options true' in the console.
if (options) {
   console.log('options true');
}
```

File: src/priority-queue.ts [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+
+		const parentIndex = Math.floor((nodeIndex - 1) / 2);
+		if (parentIndex >= this.size()) {
+			return null;
+		}
+
+		return parentIndex;
+	}
+
+	public getChildNodesIndexes(nodeIndex: number | null): [number | null, number | null] {
```
Comment (mbrich) [inline src/priority-queue.ts]: - [ ] Replace the array return with interface to clarifies how a caller uses the result.

There's no technical issues with returning an array. However, as a library used by third parties we want the signature to popup in the IDE and make usage as clear as possible. Returning a simple object isn't functionally much different than a 2 element array, but type name and property names make a huge difference in cognitive complexity.  I have to look at the function and figure out why it returns an array and how I might use that.

Here's an example type you can use:
```typescript
// this name is stupidly long
export interface ArmorPriorityQueueNodeChildren {
   left: number | null;
   right: number | null;
}
```
Then, after changing the function signature to return ArmorPriorityQueueNodeChildren<T> each array return becomes an object. You don't need to type the returns as typescript will check the type against the expected return type for you.
```typescript
// Return nulls
return { left: null, right: null };

// Return indices:
return { left: childOneIndex, right: childTwoIndex };
```

File: src/priority-queue.ts [hunk 9]
```diff
@@ -0,0 +1,262 @@
...
+		this.elements.push(element);
+		this.fixHeap(this.size() - 1);
+		return this;
+	}
+
+	public pop(): T | null {
+		if (this.size() === 0) {
+			return null;
+		}
+		if (this.size() === 1) {
```
Comment (mbrich) [inline src/priority-queue.ts:298]: My initial instinct is pop will be faster than creating and assigning a new array. Unfortunately whether it matters at all depends how frequently a single element array is popped. If you have data on this being a faster optimization we should keep it, otherwise we can simplify this short block down to:
```typescript
return this.elements.pop();
```

**Review by mbrich — COMMENTED** [review]
Didn't look at tests yet. I'll ping you to discuss after this round of changes in case test needs change.

File: src/priority-queue.ts [hunk 10]
```diff
@@ -0,0 +1,257 @@
+import {ArmorCollection} from './collection';
+import {ArmorCollectionSelector} from './selector';
+import ArmorPriorityQueueComparator from './priority-queue-comparator';
+import ArmorPriorityQueueNodeChildren from './priority-queue-children';
+import ArmorPriorityQueueOptions from './priority-queue-options';
+
+export default class ArmorPriorityQueue<T> implements ArmorCollection<T> {
+	public elements: T[];
```
Comment (mbrich) [inline src/priority-queue.ts]: - [x] Create additional ArmorPriorityQueueState<T> interface here as property with name `state`. 

- [x] Move elements property from pq to pq state class.

- [x] Add a stringify method to pq class. We'll use it to serialize queue data for storage.

- [x] Rename usages of `this.elements` to `this.state.elements`

```typescript
// Simple serialize example
public serialize(o: any): string | null {
   let result = null;
   try {
      result = JSON.stringify(this.state);
   } catch (e) {
      result = null;
   }

   return result;
}
```

- [x] Add a parse method to pq class. 
```typescript
// Simple parse example
public parse(o: string): ArmorPriorityQueueState<T> | null {
   if (typeof o !== 'string' || o === '') {
      return null;
   }
   let result: ArmorPriorityQueue<T> | null = null;
   try {
      result = JSON.parse(o);
  } catch (e) {
      result = null;
  } 

  return result;
}

Create a new interface called ArmorPriorityQueueState<T> to hold instance related data like elements. It will be used for serialization later. That way we can serialize only the queue data and not the code or other functions which won't change.
The state instance hold variables related to the queue's _current state_. State needs one variable currently: `public elements: T[];`

File: src/priority-queue.ts [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+		this.elements = [];
+		this.comparator = comparator;
+
+		if (options) {
+			this.parseOptions(options);
+		}
+	}
+
+	public parseOptions(options: ArmorPriorityQueueOptions<T>): void {
+		if (options.elements !== undefined) {
```
Comment (mbrich) [inline src/priority-queue.ts]: - [x] Check for options existence here
```typescript
if (!options) {
  return;
}
```

File: src/priority-queue.ts [hunk 12]
```diff
@@ -0,0 +1,257 @@
...
+			if (!Array.isArray(options.elements)) {
```
Comment (mbrich) [inline src/priority-queue.ts]: The above check for options.elements !== undefined can actually be eliminated due to using the isArray check. isArray will return false when options.elements is undefined or if its defined as anything else other than an array meaning both cases are handled by one check.  `options.elements` being `undefined` will not throw when called or used, but `options.elements` **will** throw when called if `options` is undefined.

Comment (mbrich) [reply src/priority-queue.ts]: Also note, you can change the isArray check to be an early drop if this function only parses `elements`.  If multiple independent properties are validated, early dropping out of one will cancel validation of all following checks which isn't necessarily what you want. In this case we would create validation functions for each property checked if any work goes into validation and call each one inside `parseOptions`. So, `parseOptionsElements`, `parseOptionsId` (made that one up), etc.

Comment (mbrich) [inline src/priority-queue.ts]: - [ ] Replace the contents of the isArray conditional with call to parse (defined in comments above).

You can gut the entire contents of everything inside the if elements is an array check, and replace it with:
```typescript
const elements = this.parse(options.elements);
if (elements) {
   this.elements = elements;
}
```

File: tests/priority-queue.spec.ts [hunk 13]
```diff
@@ -0,0 +1,599 @@
...
+	describe('size', () => {
+		it('should return 0 when priority queue is empty', () => {
+			expect(instance.size()).toBe(0);
+		});
+
+		it('should return 0 if elements are missing', () => {
+			delete instance.elements;
+			expect(instance.size()).toBe(0);
+		});
+
```
Comment (mbrich) [inline tests/priority-queue.spec.ts:224]: Add a case for count returning 1 with a single element. Sometimes weird things happen where this can be returned incorrectly.

File: tests/priority-queue.spec.ts [hunk 14]
```diff
@@ -0,0 +1,599 @@
...
+			}
+		});
+	});
+
+	describe('front', () => {
+		it('should return null when priority queue size is 0', () => {
+			expect(instance.size()).toBe(0);
+			expect(instance.front()).toBeNull();
+		});
+
```
Comment (mbrich) [inline tests/priority-queue.spec.ts:245]: Add a case for front returning the element when pq has 1 element.

File: tests/priority-queue.spec.ts [hunk 15]
```diff
@@ -0,0 +1,599 @@
...
+			expect(deepSwapped[1] === deepUnchanged[0]).toBeTruthy();
+			deepSwapped.swapNodes(1, 2);
+			expect(deepSwapped.elements[2]).toStrictEqual(deepUnchanged.elements[0]);
+			expect(deepSwapped.elements[1]).toStrictEqual(deepUnchanged.elements[2]);
+			expect(deepSwapped[2] === deepUnchanged[0]).toBeTruthy();
+			expect(deepSwapped[1] === deepUnchanged[2]).toBeTruthy();
+		});
+	});
+
+	describe('getParentNodeIndex', () => {
```
Comment (mbrich) [inline tests/priority-queue.spec.ts:342]: Add two separate test cases which check a node we'll call nodeA. 
1) Test case where you first check expected parent of nodeA, then you add a new node which changes nodeA's parent and verify the new parent is correct.
2) Test case where you first check expected parent of nodeA, then remove/pop a node which changes nodeA's parent and verify the new parent is correct.

File: tests/priority-queue.spec.ts [hunk 16]
```diff
@@ -0,0 +1,599 @@
...
+			instance.elements[0] = null!;
+			expect(instance.isHeapUnbalanced(0, 1)).toBe(false);
+			instance.elements[0] = 10;
+			instance.elements[1] = null!;
+			expect(instance.isHeapUnbalanced(0, 1)).toBe(false);
+			instance.elements[0] = null!;
+			expect(instance.isHeapUnbalanced(0, 1)).toBe(false);
+		});
+
+		it('should use the proper parameter order decided by startFromTop', () => {
```
Comment (mbrich) [inline tests/priority-queue.spec.ts]: Test is okay, but change the word _proper_ to whatever proper should confer here.

If this was written to test a strictly ascending sorted list of numbers, I might change:
`should sort elements in proper order` 
to 
`should sort elements in ascending numerical order`.
Small difference, but assume the reader won't know what proper actually means.

File: tests/priority-queue.spec.ts [hunk 17]
```diff
@@ -0,0 +1,599 @@
...
+			expect(instance.size()).toBe(limit);
+			instance.pop();
+			expect(instance.size()).toBe(limit - 1);
+		});
+
+		it('should return null when called on a priority queue of size 0', () => {
+			expect(instance.size()).toBe(0);
+			expect(instance.pop()).toBeNull();
+		});
+
```
Comment (mbrich) [inline tests/priority-queue.spec.ts:745]: Great tests here. 1 extra test case: Check that the list with 1 elements returns the element, and then null when called twice. Then one more check when list of size 5 or (some arbitrary size) is called size + 1 times, it returns n elements + 1 null.

File: tests/priority-queue.spec.ts [hunk 18]
```diff
@@ -0,0 +1,599 @@
...
+	});
+
+	describe('clear', () => {
+		it('should not throw when priority queue is empty', () => {
+			expect(instance.size()).toBe(0);
+			expect(() => {
+				instance.clear();
+			}).not.toThrow();
+		});
+
```
Comment (mbrich) [inline tests/priority-queue.spec.ts:818]: Like above, add one more case for testing the single element case.

File: src/priority-queue.ts [hunk 19]
```diff
@@ -0,0 +1,308 @@
...
+		}
+	}
+
+	public parseOptionsState(state: ArmorPriorityQueueState<T> | string): void {
+		if (!state) {
+			return;
+		}
+
+		if (typeof state === 'string') {
+			state = this.parse(state)!;
```
Comment (mbrich) [inline src/priority-queue.ts]: Line 54 is reassigning function argument `state` to the result of `this.parse(state)`.
Avoid reassigning function arguments whenever possible.
While parsing it's also a good idea to create a new result variable typically const (unless you need to modify it).
Because you're in a conditional here and need to access the result below, you can define your parse result on line 52 as follows:
```typescript
let result = ArmorPriorityQueueState<T> | null = null
```
then line 54 becomes:
```typescript
result = this.parse(state);
```
What's happening on 54 becomes more clear because we can clearly see the input and output. With the original input preserved we can also compare them.

**Review by mbrich — APPROVED** [review]
